Zend/zend_vm_def.h

```diff
@@ -8751,6 +8751,35 @@ ZEND_VM_HOT_TYPE_SPEC_HANDLER(ZEND_IS_NOT_EQUAL|ZEND_IS_NOT_IDENTICAL, (op1_info
 	ZEND_VM_SMART_BRANCH(result, 0);
 }
 
+ZEND_VM_TYPE_SPEC_HANDLER(ZEND_IS_IDENTICAL, op->op1_type == IS_CV && (op->op2_type & (IS_CONST|IS_CV)) && !(op1_info & (MAY_BE_UNDEF|MAY_BE_REF)) && !(op2_info & (MAY_BE_UNDEF|MAY_BE_REF)), ZEND_IS_IDENTICAL_NOTHROW, CV, CONST|CV, SPEC(COMMUTATIVE))
+{
+	/* This is declared below the specializations for MAY_BE_LONG/MAY_BE_DOUBLE so those will be used instead if possible. */
+	/* This optimizes $x === SOME_CONST_EXPR and $x === $y for non-refs and non-undef, which can't throw. */
+	/* (Infinite recursion when comparing arrays is an uncatchable fatal error) */
+	USE_OPLINE
+	zval *op1, *op2;
+	zend_bool result;
+
+	op1 = GET_OP1_ZVAL_PTR_UNDEF(BP_VAR_R);
+	op2 = GET_OP2_ZVAL_PTR_UNDEF(BP_VAR_R);
+	result = fast_is_identical_function(op1, op2);
```

Comment:
Can `===` generate an array recursion warning (which may be converted to exception)?

Reply:
I checked for that, and it used `_nothrow`. If the behavior changes in the future, then the opcode handler should exclude MAY_BE_ARRAY.

> `	/* (Infinite recursion when comparing arrays is an uncatchable fatal error) */` 

```php
<?php
function create_recursive_array() {
    $x = [];
    $x[0] = [&$x];
    return $x;
}
set_error_handler(function (...$args) {
    echo "In error handler\n";
});
set_exception_handler(function (...$args) {
    echo "In exception handler\n";
});
// Result: Fatal error: Nesting level too deep - recursive dependency?
var_export(create_recursive_array() === create_recursive_array());
```

Reply:
Whoops, I missed that comment. In that case this looks fine to me.